File: src/main/java/cn/labzen/web/ap/evaluate/annotation/CallEvaluator.java
package cn.labzen.web.ap.evaluate.annotation;

import cn.labzen.tool.util.Strings;
import cn.labzen.web.ap.config.Config;
import cn.labzen.web.ap.internal.Utils;
import cn.labzen.web.ap.internal.context.AnnotationProcessorContext;
import cn.labzen.web.ap.internal.element.*;
import cn.labzen.web.ap.suggestion.AppendSuggestion;
import cn.labzen.web.ap.suggestion.RemoveSuggestion;
import cn.labzen.web.ap.suggestion.ReplaceSuggestion;
import cn.labzen.web.ap.suggestion.Suggestion;
import com.google.common.collect.Lists;
import com.squareup.javapoet.TypeName;
import jakarta.annotation.Resource;

import javax.lang.model.type.TypeMirror;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import static cn.labzen.web.ap.definition.TypeNames.AP_ANNOTATION_CALL;

public final class CallEvaluator implements MethodAnnotationErasableEvaluator {

  private TypeName supportedAnnotationType;

  @Override
  public void init(AnnotationProcessorContext context) {
    supportedAnnotationType = TypeName.get(context.elements().getTypeElement(AP_ANNOTATION_CALL).asType());
  }

  @Override
  public boolean support(TypeName type) {
    return supportedAnnotationType.equals(type);
  }

  @Override
  public List<? extends Suggestion> evaluate(Config config, TypeName type, Map<String, Object> members) {
    List<Suggestion> suggestions = Lists.newArrayList(new RemoveSuggestion(Utils.getSimpleName(supportedAnnotationType), ElementMethod.class));

    Object target = members.get("target");

    String fieldName = null;
    if (target instanceof TypeMirror tm) {
      TypeName fieldClass = Utils.typeOf(tm);
      fieldName = Utils.getSimpleName(fieldClass);
      fieldName = Strings.camelCase(fieldName);

      ElementAnnotation annotation = new ElementAnnotation(TypeName.get(Resource.class));
      ElementField elementField = new ElementField(fieldName, fieldClass, List.of(annotation));
      suggestions.add(new AppendSuggestion(elementField, ElementClass.class));
    }

    Object method = members.get("method");
    if (fieldName != null || method instanceof String) {
      ElementMethodBody body = new ElementMethodBody(Strings.value(fieldName, ""), Strings.value(method, ""), Collections.emptyList());
      suggestions.add(new ReplaceSuggestion(body.keyword(), body));
    }

    return suggestions;
  }
}

File: src/main/kotlin/cn/labzen/web/ap/evaluate/annotation/CallEvaluator.kt
package cn.labzen.web.ap.evaluate.annotation

import cn.labzen.web.annotation.Call
import cn.labzen.web.ap.config.Config
import cn.labzen.web.ap.internal.Utils
import cn.labzen.web.ap.internal.element.*
import cn.labzen.web.ap.suggestion.Suggestion
import cn.labzen.web.ap.suggestion.impl.AppendSuggestion
import cn.labzen.web.ap.suggestion.impl.RemoveSuggestion
import cn.labzen.web.ap.suggestion.impl.ReplaceSuggestion
import com.squareup.javapoet.TypeName
import javax.annotation.Resource
import javax.lang.model.type.TypeMirror

class CallEvaluator : MethodErasableAnnotationEvaluator {

  override fun support(type: TypeName): Boolean =
    SUPPORTED == type

  override fun evaluate(config: Config, type: TypeName, members: Map<String, Any?>): List<Suggestion> {
    val suggestions =
      mutableListOf<Suggestion>(RemoveSuggestion(SUPPORTED_NAME, ElementMethod::class.java))

    val fieldClass = members["target"]?.let {
      Utils.typeOf(it as TypeMirror)
    }
    val fieldName = fieldClass?.let {
      val simpleName = Utils.getSimpleName(it)
      simpleName[0].lowercaseChar() + simpleName.substring(1)
    }
    fieldName?.run {
      val annotation = ElementAnnotation(TypeName.get(Resource::class.java))
      val field = ElementField(fieldName, fieldClass, listOf(annotation))
      suggestions.add(AppendSuggestion(field, ElementClass::class.java))
    }

    val methodName = members["method"] as String?
    if (fieldName != null || methodName != null) {
      val body = ElementMethodBody(fieldName ?: "", methodName ?: "", emptyList())
      suggestions.add(ReplaceSuggestion(body.keyword(), body))
    }

    return suggestions
  }

  companion object {
    private val SUPPORTED = TypeName.get(Call::class.java)
    private val SUPPORTED_NAME = Utils.getSimpleName(SUPPORTED)
  }
}